From 3cbc8ab5b6308debb01cdfd997565a4397572014 Mon Sep 17 00:00:00 2001
From: Aaron <kljsandjb@me.com>
Date: Fri, 11 Sep 2020 13:25:55 +0000
Subject: [PATCH] LAB 5 challenge 1, IRQ driven disk access

finished challenge 1

make[1]: Leaving directory '/usr/src/app/lab'
internal FS tests [fs/test.c]: OK (29.2s)
  fs i/o: OK
  check_bc: OK
  check_super: OK
  check_bitmap: OK
  alloc_block: OK
  file_open: OK
  file_get_block: OK
  file_flush/file_truncate/file rewrite: OK
testfile: OK (25.7s)
  serve_open/file_stat/file_close: OK
  file_read: OK
  file_write: OK
  file_read after file_write: OK
  open: OK
  large file: OK
spawn via spawnhello: OK (25.0s)
Protection I/O space: OK (25.4s)
    (Old jos.out.faultio failure log removed)
PTE_SHARE [testpteshare]: OK (24.9s)
PTE_SHARE [testfdsharing]: OK (25.0s)
start the shell [icode]: Timeout! OK (51.3s)
testshell: OK (29.1s)
    (Old jos.out.testshell failure log removed)
primespipe: OK (34.3s)
Score: 150/150
root@c9982579a1dc:/usr/src/app/lab#
---
 fs/ide.c       | 54 +++++++++++++++++++++++++++++---------------------
 inc/env.h      |  7 ++++++-
 inc/lib.h      |  1 +
 inc/syscall.h  |  1 +
 kern/picirq.c  | 10 ++++++++--
 kern/sched.c   |  2 +-
 kern/syscall.c | 32 ++++++++++++++++++++++++++++++
 kern/trap.c    | 34 ++++++++++++++++++++++++++++++-
 lib/fork.c     |  6 +++++-
 lib/syscall.c  |  5 +++++
 10 files changed, 123 insertions(+), 29 deletions(-)

diff --git a/fs/ide.c b/fs/ide.c
index 2d8b4bf..21ef938 100644
--- a/fs/ide.c
+++ b/fs/ide.c
@@ -59,7 +59,6 @@ ide_set_disk(int d)
 	diskno = d;
 }
 
-
 int
 ide_read(uint32_t secno, void *dst, size_t nsecs)
 {
@@ -69,18 +68,17 @@ ide_read(uint32_t secno, void *dst, size_t nsecs)
 
 	ide_wait_ready(0);
 
-	outb(0x1F2, nsecs);
-	outb(0x1F3, secno & 0xFF);
-	outb(0x1F4, (secno >> 8) & 0xFF);
-	outb(0x1F5, (secno >> 16) & 0xFF);
-	outb(0x1F6, 0xE0 | ((diskno&1)<<4) | ((secno>>24)&0x0F));
-	outb(0x1F7, 0x20);	// CMD 0x20 means read sector
+	// outb(0x1F2, nsecs);
+	// outb(0x1F3, secno & 0xFF);
+	// outb(0x1F4, (secno >> 8) & 0xFF);
+	// outb(0x1F5, (secno >> 16) & 0xFF);
+	// outb(0x1F6, 0xE0 | ((diskno&1)<<4) | ((secno>>24)&0x0F));
+	// outb(0x1F7, nsecs > 1 ? 0xc4 : 0x20);	// CMD 0x20 means read sector
 
-	for (; nsecs > 0; nsecs--, dst += SECTSIZE) {
-		if ((r = ide_wait_ready(1)) < 0)
-			return r;
-		insl(0x1F0, dst, SECTSIZE/4);
-	}
+	// between issuing disk cmd and set to sleep, there might be a timer IRQ
+	// comes in, and fs -> RUNNABLE, then disk IRQ comes, and we handle it
+	// then fs goes to sleep, with CPU halted, we missed wake up
+	sys_ide_sleep(secno, dst, nsecs, 0, diskno);
 
 	return 0;
 }
@@ -94,19 +92,29 @@ ide_write(uint32_t secno, const void *src, size_t nsecs)
 
 	ide_wait_ready(0);
 
-	outb(0x1F2, nsecs);
-	outb(0x1F3, secno & 0xFF);
-	outb(0x1F4, (secno >> 8) & 0xFF);
-	outb(0x1F5, (secno >> 16) & 0xFF);
-	outb(0x1F6, 0xE0 | ((diskno&1)<<4) | ((secno>>24)&0x0F));
-	outb(0x1F7, 0x30);	// CMD 0x30 means write sector
+	// outb(0x1F2, nsecs);
+	// outb(0x1F3, secno & 0xFF);
+	// outb(0x1F4, (secno >> 8) & 0xFF);
+	// outb(0x1F5, (secno >> 16) & 0xFF);
+	// outb(0x1F6, 0xE0 | ((diskno&1)<<4) | ((secno>>24)&0x0F));
+	// outb(0x1F7, nsecs > 1 ? 0xc5 : 0x30);	// CMD 0x30 means write sector
 
-	for (; nsecs > 0; nsecs--, src += SECTSIZE) {
-		if ((r = ide_wait_ready(1)) < 0)
-			return r;
-		outsl(0x1F0, src, SECTSIZE/4);
-	}
+	// outsl(0x1F0, src, nsecs*SECTSIZE/4);
+	sys_ide_sleep(secno, (void *)src, nsecs, 1, diskno);
 
 	return 0;
 }
 
+/* 
+
+0. non-fs env raises a fs request and sleep
+1. fs might issue read/write command
+2. fs goes to sleep and yields the CPU
+3. disk irq comes, fs wakes up
+4. fs in turn wakes up env that waits
+5. fs serv goes for another run
+
+So basically we can only handle one IRQ each
+time (since fs env will sleep after each disk command),
+we can support concurrency by implementing 'thread' in JOS
+*/
\ No newline at end of file
diff --git a/inc/env.h b/inc/env.h
index ab392db..c8be428 100644
--- a/inc/env.h
+++ b/inc/env.h
@@ -35,7 +35,8 @@ enum {
 	ENV_DYING,
 	ENV_RUNNABLE,
 	ENV_RUNNING,
-	ENV_NOT_RUNNABLE
+	ENV_NOT_RUNNABLE,
+	ENV_IDE_SLEEPING
 };
 
 // Special environment types
@@ -66,6 +67,10 @@ struct Env {
 	uint32_t env_ipc_value;		// Data value sent to us
 	envid_t env_ipc_from;		// envid of the sender
 	int env_ipc_perm;		// Perm of page mapping received
+
+	// Lab 5 FS
+	void *chan;				// sleep on channel (0 means write, otherwise read)
+	int op;				// read 0, write 1
 };
 
 #endif // !JOS_INC_ENV_H
diff --git a/inc/lib.h b/inc/lib.h
index 88f7102..089ea85 100644
--- a/inc/lib.h
+++ b/inc/lib.h
@@ -57,6 +57,7 @@ int	sys_page_map(envid_t src_env, void *src_pg,
 int	sys_page_unmap(envid_t env, void *pg);
 int	sys_ipc_try_send(envid_t to_env, uint32_t value, void *pg, int perm);
 int	sys_ipc_recv(void *rcv_pg);
+void sys_ide_sleep(uint32_t secno, void *chan, size_t nsecs, int op, int diskno);
 
 // This must be inlined.  Exercise for reader: why?
 static inline envid_t __attribute__((always_inline))
diff --git a/inc/syscall.h b/inc/syscall.h
index 20c6433..cc21b42 100644
--- a/inc/syscall.h
+++ b/inc/syscall.h
@@ -17,6 +17,7 @@ enum {
 	SYS_yield,
 	SYS_ipc_try_send,
 	SYS_ipc_recv,
+	SYS_ide_sleep,
 	NSYSCALLS
 };
 
diff --git a/kern/picirq.c b/kern/picirq.c
index 8cb3e62..adc9080 100644
--- a/kern/picirq.c
+++ b/kern/picirq.c
@@ -11,10 +11,16 @@
 uint16_t irq_mask_8259A = 0xFFFF & ~(1<<IRQ_SLAVE);
 static bool didinit;
 
-/* Initialize the 8259A interrupt controllers. */
+/* Initialize the 8259A interrupt controllers.
+ * https://k.lse.epita.fr/internals/8259a_controller.html
+*/
 void
 pic_init(void)
 {
+	// init ide disk interrupt
+	// for simplicity we init ide's irq just before pic init
+	// otherwise we might do it in IOAPIC
+	irq_setmask_8259A(irq_mask_8259A & ~(1<<IRQ_IDE));
 	didinit = 1;
 
 	// mask all interrupts
@@ -48,7 +54,7 @@ pic_init(void)
 
 	// Set up slave (8259A-2)
 	outb(IO_PIC2, 0x11);			// ICW1
-	outb(IO_PIC2+1, IRQ_OFFSET + 8);	// ICW2
+	outb(IO_PIC2+1, IRQ_OFFSET + 8);	// ICW2	base = 40, IDE will be taken care from slave
 	outb(IO_PIC2+1, IRQ_SLAVE);		// ICW3
 	// NB Automatic EOI mode doesn't tend to work on the slave.
 	// Linux source code says it's "to be investigated".
diff --git a/kern/sched.c b/kern/sched.c
index b9e8da9..7cd8161 100644
--- a/kern/sched.c
+++ b/kern/sched.c
@@ -64,7 +64,7 @@ sched_halt(void)
 	for (i = 0; i < NENV; i++) {
 		if ((envs[i].env_status == ENV_RUNNABLE ||
 		     envs[i].env_status == ENV_RUNNING ||
-		     envs[i].env_status == ENV_DYING))
+		     envs[i].env_status == ENV_DYING || envs[i].env_status == ENV_IDE_SLEEPING))
 			break;
 	}
 	if (i == NENV) {
diff --git a/kern/syscall.c b/kern/syscall.c
index 315f404..311632a 100644
--- a/kern/syscall.c
+++ b/kern/syscall.c
@@ -385,6 +385,36 @@ sys_ipc_recv(void *dstva)
 	return 0;
 }
 
+static void
+sys_ide_sleep(size_t secno, void *chan, size_t nsecs, int op, int diskno)
+{
+	int r;
+	if (op == 0)
+	{
+		outb(0x1F2, nsecs);
+		outb(0x1F3, secno & 0xFF);
+		outb(0x1F4, (secno >> 8) & 0xFF);
+		outb(0x1F5, (secno >> 16) & 0xFF);
+		outb(0x1F6, 0xE0 | ((diskno&1)<<4) | ((secno>>24)&0x0F));
+		outb(0x1F7, nsecs > 1 ? 0xc4 : 0x20);	// CMD 0x20 means read sector
+	}
+	else
+	{
+		outb(0x1F2, nsecs);
+		outb(0x1F3, secno & 0xFF);
+		outb(0x1F4, (secno >> 8) & 0xFF);
+		outb(0x1F5, (secno >> 16) & 0xFF);
+		outb(0x1F6, 0xE0 | ((diskno&1)<<4) | ((secno>>24)&0x0F));
+		outb(0x1F7, nsecs > 1 ? 0xc5 : 0x30);	// CMD 0x30 means write sector
+
+		outsl(0x1F0, chan, nsecs*512/4);
+	}
+    curenv->chan = chan;
+	curenv->env_status = ENV_IDE_SLEEPING;
+	curenv->op = op;
+	sched_yield();
+}
+
 // Dispatches to the correct kernel function, passing the arguments.
 int32_t
 syscall(uint32_t syscallno, uint32_t a1, uint32_t a2, uint32_t a3, uint32_t a4, uint32_t a5)
@@ -425,6 +455,8 @@ syscall(uint32_t syscallno, uint32_t a1, uint32_t a2, uint32_t a3, uint32_t a4,
 		return sys_ipc_try_send(a1, a2, (void *)a3, a4);
 	case SYS_ipc_recv:
 		return sys_ipc_recv((void *)a1);
+	case SYS_ide_sleep:
+		sys_ide_sleep(a1, (void *)a2, a3, (int)a4, (int)a5);
 	case NSYSCALLS:
 	default:
 		return -E_INVAL;
diff --git a/kern/trap.c b/kern/trap.c
index ccf82f7..78c732e 100644
--- a/kern/trap.c
+++ b/kern/trap.c
@@ -238,20 +238,51 @@ trap_dispatch(struct Trapframe *tf)
 	// LAB 4: Your code here.
 	if (tf->tf_trapno >= IRQ_OFFSET && tf->tf_trapno < IRQ_OFFSET + 16)
 	{
-		lapic_eoi();
 		switch (tf->tf_trapno)
 		{
 		case IRQ_OFFSET + IRQ_TIMER:
+			lapic_eoi();
 			sched_yield();
 			break;
 
 		case IRQ_OFFSET + IRQ_KBD:
 			kbd_intr();
+			lapic_eoi();
 			return;
 
 		case IRQ_OFFSET + IRQ_SERIAL:
 			serial_intr();
+			lapic_eoi();
 			return;
+
+		case IRQ_OFFSET + IRQ_IDE:
+			for (int i = 0; i < NENV; i++)
+			{
+				if (envs[i].env_type == ENV_TYPE_FS)
+				{
+					// read a BLKSIZE from disk if needed then acknowledge the interrupt
+					if (envs[i].op == 0)
+					{
+						lcr3(PADDR(envs[i].env_pgdir));
+						insl(0x1F0, envs[i].chan, PGSIZE / 4);
+						envs[i].chan = 0;
+						lcr3(PADDR(kern_pgdir));
+					}
+					// OCW2: send non-specific EOI command to give driver an ACK
+					// otherwise we won't receive the rest IDE interrupts followed
+					outb(IO_PIC1, 0x20);
+					outb(IO_PIC2, 0x20);
+					// finally, make fs runnable
+					if (envs[i].env_status == ENV_IDE_SLEEPING)
+						envs[i].env_status = ENV_RUNNABLE;
+					else
+					{
+						cprintf("status: %u\n", envs[i].env_status);
+						print_trapframe(tf);
+					}
+					return;
+				}
+			}
 		
 		default:
 			break;
@@ -427,6 +458,7 @@ page_fault_handler(struct Trapframe *tf)
 	utf->utf_eflags = tf->tf_eflags;
 	utf->utf_esp = tf->tf_esp;
 
+	// page fault exception handler thread
 	tf->tf_esp = (uintptr_t)&utf->utf_fault_va;
 	tf->tf_eip = (uintptr_t)curenv->env_pgfault_upcall;
 	env_run(curenv);
diff --git a/lib/fork.c b/lib/fork.c
index ce606d6..4cdf73c 100644
--- a/lib/fork.c
+++ b/lib/fork.c
@@ -63,6 +63,11 @@ pgfault(struct UTrapframe *utf)
 // copy-on-write again if it was already copy-on-write at the beginning of
 // this function?)
 //
+// Answer of above exercise question:
+// *If the beginning it is CoW, and then it faults, marked as PTE_W, then
+// we'll corrupt the child's copy when this page is being written by parent
+// after this, and the 'snapshot' misfunction*
+//
 // Returns: 0 on success, < 0 on error.
 // It is also OK to panic on error.
 //
@@ -73,7 +78,6 @@ duppage(envid_t envid, unsigned pn)
 
 	// LAB 4: Your code here.
 	int perm = PTE_P | PTE_U;	// at least PTE_P and PTE_U
-	// envid_t curenvid = sys_getenvid();
 
 	int is_wr = (uvpt[pn] & PTE_W) == PTE_W;
 	int is_cow = (uvpt[pn] & PTE_COW) == PTE_COW;
diff --git a/lib/syscall.c b/lib/syscall.c
index f6d1136..c033c9d 100644
--- a/lib/syscall.c
+++ b/lib/syscall.c
@@ -158,3 +158,8 @@ sys_ipc_recv(void *dstva)
 	return syscall(SYS_ipc_recv, 1, (uint32_t)dstva, 0, 0, 0, 0);
 }
 
+void
+sys_ide_sleep(uint32_t secno, void *chan, size_t nsecs, int op, int diskno)
+{
+	syscall(SYS_ide_sleep, 0, secno, (uint32_t)chan, nsecs, (uint32_t)op, (uint32_t)diskno);
+}
\ No newline at end of file
-- 
2.20.1

